docs/proposals/generalize-modules.md

```diff
@@ -0,0 +1,72 @@
+---
+title: "Generalize Modules Service to make it extensible"
+linkTitle: "Generalize Modules Service to make it extensible"
+weight: 1
+slug: generalize-modules
+---
+
+- Author: @annanay25
+- Reviewers: 
+- Date: April 2020
+- Status: Draft
+
+## Overview
+
+Cortex uses modules to start and operate services with dependencies. Inter-service dependencies are specified in a map and passed to a module manager which ensures that they are started in the right order of dependencies. While this works really well, the implementation is tied in specifically to the Cortex struct and is not flexible for use with other projects like Loki, which also require similar forms of dependency management. 
+ 
+We would like to extend modules in cortex to a generic dependency management framework, that can be used by any project with no ties to cortex. 
+
+## Specific goals
+
+- Framework should allow for reusing cortex modules and allow us to:
+  - Add new modules
+  - Overwrite the implementation of a current module
+  - Manage dependencies
+- Framework should allow for building an application from scratch using the `modules` package, with no dependencies on Cortex. For ex: Remove code from Loki that was copied from `pkg/cortex/cortex.go`.
+
+
+
+## Proposed Design
+
+To make the modules package extensible, we need to abstract away any Cortex specific details from the module manager. The proposed design is to -
+
+- Make a new component `moduleManager`, which is envisioned to be a central manager for all modules of the application. It stores modules & dependencies, and will be housed under a new package `pkg/util/modules`. The following is the interface for interacting with the `moduleManager`:
+```
+// Manager defines the interface for interaction with moduleManager.
+type Manager interface {
+   RegisterModule(name string, deps []string, svc service, options ...Option)
```

Comment:
What is `service` here? Module needs to pass a function that creates a service (like various `init` functions in Cortex).

Reply:
`service` is a `func()` type, would it be better renamed as `initFn`?

Reply:
`service` type was unclear to me. Yes, I think `initFn func() (Service, error)` would be more descriptive. It should be documented that it can return nil (right? that's what current init functions do), in which case no service is returned for this module ... but if any other service depends on it, how should that behave?

Reply:
> It should be documented that it can return nil

Yes. I'll make sure to add that.

> how should that behave?

In `InitModuleServices` we can error out if `initFn()` returns `nil` and the module has "inverse dependencies".